File: python/day2/aoc2.py
with open("/Users/christopherauld/Desktop/AOC2018/data/day2/input.txt", "r") as file:
    s = [line.strip("\n") for line in file]

# Part 1
out = {}

for group in s:
    for char in group:
        if group.count(char) > 1:
            out[group] = [0, 0]
        if group.count(char) == 2:
            out[group][0] = 1
        if group.count(char) == 3:
            out[group][1] = 1

twos = 0
threes = 0
for key, val in out.items():
    twos += val[0]
    threes += val[1]
print(twos * threes)

# Part 2
o = {}
for i, group in enumerate(s):
    o[i] = group

def is_off_by_one(a, b):
    num_wrong = 0
    for i, char in enumerate(a):
        if char != b[i]:
            num_wrong += 1
        if num_wrong > 1:
            return False
    if num_wrong == 1:
        return True
    if num_wrong != 1:
        return False

for key, val in o.items():
    for k, v in o.items():
        found = is_off_by_one(val, v)
        if found:
            print(k, v, key, val)

File: go/day2/aoc2.go
package main

import (
    "fmt"
    "os"
    "bufio"
    "strings"
)

func IsOffByOne(a string, b string) bool {
    num_wrong := 0
    for i := range a {
        a_char := string(a[i])
        b_char := string(b[i])
        if a_char != b_char {
            num_wrong++
        }
        if num_wrong > 1 {
            return false
        }
    }
    if num_wrong == 1 {
        return true
    } else {
        return false
    }
}

func main() {
    filename := "/Users/christopherauld/Desktop/AOC2018/data/day2/input.txt"
    b, err := os.Open(filename)
    if err != nil {
        fmt.Println(err)
    }

    // Part 1
    scanner := bufio.NewScanner(b)
    twos := 0
    threes := 0
    labels := make(map[int]string)
    i := 0

    for scanner.Scan() {
        found_two := false
        found_three := false
        label := scanner.Text()
        labels[i] = label
        i++
        for idx := range label{
            if strings.Count(label, string(label[idx])) == 2  && !found_two {
                twos++
                found_two = true
            } 
            if strings.Count(label, string(label[idx])) == 3  && !found_three {
                threes++ 
                found_three = true
            }
        }
    }
    fmt.Println(twos * threes)

    // Part 2
    for i = range labels {
        for j := range labels {
            if IsOffByOne(labels[i], labels[j]) {
                fmt.Println(labels[i], labels[j])
            }
        }
    }
}